cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java

```diff
@@ -835,327 +766,261 @@ public Service generateService() {
+            Service service = createService(
+                    serviceName,
+                    ListenersUtils.getServiceType(listener),
+                    ports,
+                    getLabelsWithStrimziName(name, templatePerPodServiceLabels),
+                    selector,
+                    Util.mergeLabelsOrAnnotations(ListenersUtils.brokerDnsAnnotations(listener, pod), templatePerPodServiceAnnotations)
+            );
+
+            if (KafkaListenerType.LOADBALANCER == listener.getType()) {
+                String loadBalancerIP = ListenersUtils.brokerLoadBalancerIP(listener, pod);
+                if (loadBalancerIP != null) {
+                    service.getSpec().setLoadBalancerIP(loadBalancerIP);
+                }
+            }
 
-            if (isExposedWithLoadBalancer()) {
-                if (templatePerPodServiceLoadBalancerSourceRanges != null) {
+            if (KafkaListenerType.LOADBALANCER == listener.getType()) {
+                List<String> loadBalancerSourceRanges = ListenersUtils.loadBalancerSourceRanges(listener);
+                if (loadBalancerSourceRanges != null) {
+                    service.getSpec().setLoadBalancerSourceRanges(loadBalancerSourceRanges);
+                } else if (templatePerPodServiceLoadBalancerSourceRanges != null) {
                     service.getSpec().setLoadBalancerSourceRanges(templatePerPodServiceLoadBalancerSourceRanges);
                 }
             }
 
-            if (isExposedWithLoadBalancer() || isExposedWithNodePort()) {
-                if (templatePerPodServiceTrafficPolicy != null)  {
+            if (KafkaListenerType.LOADBALANCER == listener.getType() || KafkaListenerType.NODEPORT == listener.getType()) {
+                ExternalTrafficPolicy etp = ListenersUtils.externalTrafficPolicy(listener);
+                if (etp != null) {
+                    service.getSpec().setExternalTrafficPolicy(etp.toValue());
+                } else if (templatePerPodServiceTrafficPolicy != null) {
                     service.getSpec().setExternalTrafficPolicy(templatePerPodServiceTrafficPolicy.toValue());
                 }
             }
 
-            return service;
+            services.add(service);
         }
 
-        return null;
+        return services;
     }
 
-    /**
-     * Generates route for pod. This route is used for exposing it externally using OpenShift Routes.
+        /**
+     * Generates a list of bootstrap route which can be used to bootstrap clients outside of OpenShift.
      *
-     * @param pod Number of the pod for which this route should be generated
-     * @return The generated Route
+     * @return The list of generated Routes
      */
-    public Route generateExternalRoute(int pod) {
-        if (isExposedWithRoute()) {
-            String perPodServiceName = externalServiceName(cluster, pod);
+    public List<Route> generateExternalBootstrapRoutes() {
```

Comment:
Does this not break us for non openshift?
Since it is always deploying with routes?

Reply:
In a short, it does not. As a longer answer, this just generates some objects when called from `KafkaAssemblyOperator` which decides when to call it and when to not all it. So the protection is in there. I develop on Kubernetes, so this does not cause any issues.